web_poet/page_inputs/client.py

```diff
@@ -39,8 +40,20 @@ class HttpClient:
     :ref:`advanced-downloader-impl` documentation.
     """
 
-    def __init__(self, request_downloader: Callable = None):
+    def __init__(
+        self,
+        request_downloader: Callable = None,
+        *,
+        save_responses: bool = False,
+        return_only_saved_responses: bool = False,
+        responses: Optional[Iterable[Tuple[HttpRequest, HttpResponse]]] = None,
+    ):
         self._request_downloader = request_downloader or _perform_request
+        self.save_responses = save_responses
+        self.return_only_saved_responses = return_only_saved_responses
+        self._saved_responses: Dict[str, Tuple[HttpRequest, HttpResponse]] = {
+            request_fingerprint(req): (req, resp) for req, resp in responses or {}
+        }
```

Comment:
I don't like this (and the related code in other places) that much, but the alternative is to calculate fingerprints on each comparison which doesn't look like a good idea to me, and we can't throw out requests because we need to serialize them. Maybe there is a better way to store this.

Reply:
There are alternatives like using a cache for the fingerprinting function, but it looks fine as-is to me.